cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,544 @@
+   * @throws IOException
+   */
+  private SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(metaKey)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(metaKey, password);
+      return ((KeyMetadata) key).metadata;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the securely stored data as a UTF8 encoded byte array.
+   * @param name Name of the element in the secure store
+   * @return The data associated with the element as an UTF8 formatted byte array
+   * @throws IOException
+   */
+  private byte[] getData(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      SecretKeySpec key = (SecretKeySpec) keyStore.getKey(name, password);
+      return key.getEncoded();
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the key " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  private static Path constructOldPath(Path path) {
+    return Paths.get(path.toString(), "_OLD");
+  }
+
+  private static Path constructNewPath(Path path) {
+    return Paths.get(path.toString(), "_NEW");
+  }
+
+  private static void loadFromPath(KeyStore keyStore, Path path, char[] password)
+    throws IOException {
+    try (InputStream in = new DataInputStream(Files.newInputStream(path))) {
+      keyStore.load(in, password);
+    } catch (NoSuchAlgorithmException | CertificateException e) {
+      throw new IOException("Unable to load the Secure Store. ", e);
+    }
+  }
+
+  private static boolean isBadOrWrongPassword(IOException ioe) {
+    /* According to Java keystore documentation if the load failed due to bad password
+       then the cause of the exception would be set to "UnrecoverableKeyException".
+       Unfortunately that is not the observed behavior. */
+    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+      return true;
+    }
+    // Workaround
+    return (ioe.getCause() == null)
+      && (ioe.getMessage() != null)
+      && ((ioe.getMessage().contains("Keystore was tampered")) || (ioe
+      .getMessage().contains("password was incorrect")));
+  }
+
+  private static String constructMetadataKey(String name) {
+    return name + METADATA_SUFFIX;
+  }
+
+  /**
+   * Initialize the keyStore.
+   *
+   * @throws IOException If there is a problem reading or creating the keystore.
+   */
+  private void locateKeystore() throws IOException {
```

Comment:
Since this method is only called from constructor, better make this method static to avoid fields being used directly.